cargo-screeps/src/upload.rs

```diff
@@ -94,3 +92,22 @@ pub fn upload(root: &Path, config: &Configuration) -> Result<(), failure::Error>
 
     Ok(())
 }
+
+fn authorize(
+    request: reqwest::RequestBuilder,
+    upload_config: &UploadConfiguration,
+) -> Result<reqwest::RequestBuilder, failure::Error> {
+    if upload_config.auth_token.is_some() {
+        Ok(request.header(
+            "X-Token",
+            upload_config.auth_token.as_ref().unwrap().as_str(),
+        ))
+    } else if upload_config.username.is_some() && upload_config.password.is_some() {
+        Ok(request.basic_auth(
+            upload_config.username.as_ref().unwrap(),
+            upload_config.password.as_ref(),
+        ))
+    } else {
+        bail!("could not authorize request. must set auth_token or username/password in [upload] config");
```

Comment:
Ideally `UploadConfiguration` should always represent a valid configuration, and we shouldn't have to bail out later because of invalid configuration options.

Do you think we could bail out here instead?

 https://github.com/daboross/screeps-in-rust-via-wasm/blob/430c097ea9d8340dfa2aa6d2fe26c475af949c85/cargo-screeps/src/config.rs#L110-L134

We could then represent the authentication with something like

```rust
pub struct Authentication {
    Token(String),
    UserPass { username: String, password: String },
}
```

Or just leave it as is, and make this bail an `unreachable!()` panic.

Reply:
Makes sense! I'll see what I can do.